xdsl/dialects/accfg.py

```diff
@@ -0,0 +1,310 @@
+from __future__ import annotations
+
+from collections.abc import Iterable, Sequence
+
+from xdsl.dialects.builtin import (
+    AnyIntegerAttr,
+    ArrayAttr,
+    DictionaryAttr,
+    IntegerAttr,
+    StringAttr,
+    SymbolRefAttr,
+    i32,
+)
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    Operation,
+    ParametrizedAttribute,
+    SSAValue,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    AttrSizedOperandSegments,
+    IRDLOperation,
+    ParameterDef,
+    VerifyException,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_operand_def,
+    prop_def,
+    result_def,
+    var_operand_def,
+)
+from xdsl.traits import SymbolOpInterface
+
+
+@irdl_attr_definition
+class TokenType(ParametrizedAttribute, TypeAttribute):
+    """
+    Async token type for managing synchronization between asynchronously
+    launched accelerators. You can use them to create def-use chains
+    between accfg.launch and accfg.await ops.
+    """
+
+    name = "accfg.token"
+
+    accelerator: ParameterDef[StringAttr]
+
+    def __init__(self, accelerator: str | StringAttr):
+        if not isinstance(accelerator, StringAttr):
+            accelerator = StringAttr(accelerator)
+        return super().__init__([accelerator])
+
+
+@irdl_attr_definition
+class StateType(ParametrizedAttribute, TypeAttribute):
+    """
+    State type to manage accelerator configuration state.
+    States are chained through the def-use chain to allow
+    deduplication of setup calls.
+    """
+
+    name = "accfg.state"
+
+    accelerator: ParameterDef[StringAttr]
+
+    def __init__(self, accelerator: str | StringAttr):
+        if not isinstance(accelerator, StringAttr):
+            accelerator = StringAttr(accelerator)
+        return super().__init__([accelerator])
+
+
+class AcceleratorSymbolOpTrait(SymbolOpInterface):
+    """
+    Symbol op trait to define multiple accelerator names.
+    """
+
+    def get_sym_attr_name(self, op: Operation) -> StringAttr | None:
+        assert isinstance(op, AcceleratorOp)
+        return StringAttr(op.name_prop.string_value())
+
+
+@irdl_op_definition
+class AcceleratorOp(IRDLOperation):
+    """
+    Declares an accelerator that can be configured, launched, etc.
+    `fields` is a dictionary mapping accelerator configuration names to
+    registers addresses.
+    """
+
+    name = "accfg.accelerator"
+
+    traits = frozenset([AcceleratorSymbolOpTrait()])
+
+    name_prop = prop_def(SymbolRefAttr, prop_name="name")
+
+    fields = prop_def(DictionaryAttr)
+
+    launch_addr = prop_def(IntegerAttr)
+
+    barrier = prop_def(IntegerAttr)  # TODO: this will be reworked in a later version
```

Comment:
What do you mean by reworked?

Reply:
The exact translation of `accfg.await` is not final, and as such the information required for this on the `accfg.accelerator` op will probably change again in the future. We're looking into ways of generalizing this aspect currently, but this is a thing that actually works for snitch now.

Reply:
it might be worth adding this to the comment

Reply:
Will do!